### ISSUE TITLE
x/tools/godoc: unexported fields are displayed in exported variable initializers

### ISSUE THREAD

**Issue by opennota** [golang/go#22803]
### What did you do?

https://godoc.org/github.com/opennota/rwc

### What did you expect to see?

godoc doesn't show unexported fields.

### What did you see instead?

godoc for the `DefaultConstructor` variable shows unexported fields:

```
var DefaultConstructor = Constructor{
    ng4: [32768]uint32{ /* 32768 elements not displayed */

    },
    ng3: [1024]uint32{ /* 1024 elements not displayed */

    },
    ng3beg: [1024]uint32{ /* 1024 elements not displayed */

    },
    ng3end: [1024]uint32{ /* 1024 elements not displayed */

    },
    ng2: [32]uint32{
        18874896, 134742016, 134234113, 16424, 536887585, 1073748512,
        289, 16417, 2101380, 0, 679937, 2416460032, 134217729, 540961,
        2372130, 16385, 1, 1610629156, 1208500513, 3280960, 524289, 16384,
        0, 1056, 1, 0, 0, 4096, 0, 2097664, 8, 34320,
    },
    ng1: 2148156709,
}
```

**Comment by bradfitz** [comment]
@robpike, @griesemer, any opinions? Is this expected?

**Comment by griesemer** [comment]
I think this is expected (I don't think we have special code to filter those fields, but I haven't checked); when there's an initialization expression we simply print that expression. That is not to say, this is correct (or wrong for that matter).

**Comment by robpike** [comment]
It doesn't bother me much, although I can understand why some would be distracted.

**Comment by rsc** [comment]
The proposal-review group is fine with unexported field initializations being replaced with "..." just as for unexported fields in type definitions. We don't anticipate working on it ourselves.

**Comment by agnivade** [comment]
I am thinking of working on this. This is my first time tinkering with ASTs, so I would need a little guidance.

Just to be clear, the fix needs to be done in the `go/ast` package which actually generates printed representation of the AST, and not inside `x/tools/godoc`.

Initially, I started making changes in the `go/printer` package, but then later realized that the filtering of fields and idents is done from inside the `go/ast` package. So, I have started work there. I have a working version ready which needs some finishing touches. Here is what I have done so far:

- Added an `Incomplete` field inside `CompositeLit` similar to `StructType` to signal the printer package that fields are missing.
- Added a `filterExpressionList` function in `ast/filter.go` which goes through a slice of `[]Expr` checking for KeyValue expressions inside CompositeLiterals. If it finds that the Ident.Name of the Key is not exported, it filters that element from the slice.
- Pending is to take this information in the `go/printer` package and either print a `...` or the usual `// contains filtered or unexported fields`.

Does this look like the correct approach ?
